Attitude Controller

shihaocao opened the pull request:
# Attitude Controller

Fixes #287 

### Summary of changes
- Removed AttitudeComputer from MCL
- Added AttitudeController (does not need temporary data structures for mtr and rwa_torques)
- Added AttitudeController pio unit test
- Bump psim and lin submodule pointers
- ADCSCommander no longer needs to read pointing vectors, instead reads suggested mtr and rwa_torques
- AttitudeComputer replaced by AttitudeController in FSW, MCL
- ADCSCommander is restructured to hopefully be simpler
- ADCSCommander has upgraded pio unit test
- New differentialation of adcs_state's:
- `point_manual` is used by the flight computer/gs in the case where we set the pointing vectors from the ground
- `manual` is what should be used by ptest cases that directly set the actuators!
### Testing
- Deferring exhaustive attitude controller testing to later intergated ptest + psim testing
- Basic function pio unit test to confirm that AttitudeController won't brick FSW
- pio unit tested the new ADCSCommander

### Constants
N/A

### Telemetry
- AttitudeComputer replaced by AttitudeController in relevant fields
- renamed pointing objective fields
- corrected the proper critical commands in Core ADCS telemetry

Also, take the opportunity to add the telemetry into the appropriate flows in src/flow_data.csv.
- Yes I did the above, it hurt me though lol

### Documentation Evidence
- On the way...

kkrol27 commented on src/.~lock.flow_data.csv#:
```diff
@@ -0,0 +1 @@
+,shihao,shihao-T490V2,18.10.2020 02:03,file:///home/shihao/.config/libreoffice/4;
```
Not sure what this is or if it's supposed to be in here?

kkrol27 replied:
Not sure what this is or if it needs to be in here?

kkrol27 commented on src/fsw/FCCode/AttitudeController.hpp:
```diff
@@ -0,0 +1,120 @@
+     * @param offset
+     */
+    AttitudeController(StateFieldRegistry &registry, unsigned int offset);
+    ~AttitudeController() = default; //is this needed?
```
There only really a point in specifying this if we mark it as `virtual`. However, given we never deconstruct control tasks it doesn't really matter lmao.
```suggestion
```

kkrol27 commented on src/fsw/FCCode/AttitudeController.hpp:
```diff
@@ -0,0 +1,120 @@
+    void execute() override;
+
+   protected:
+    // Inputs from the adcs box monitor
+    ReadableStateField<lin::Vector3f> *const w_wheels_rd_fp;
```
Don't we need to read in the magnetic field from here as well? For the detumbler that is?

kkrol27 commented on src/fsw/FCCode/AttitudeController.hpp:
```diff
@@ -0,0 +1,120 @@
+    WritableStateField<lin::Vector3f> pointer_vec1_desired_f;
+    WritableStateField<lin::Vector3f> pointer_vec2_desired_f;
+
+    // Internal to this statefield, data containers
```
```suggestion
    // Internal to this control task, data containers
```

kkrol27 commented on src/fsw/FCCode/AttitudeController.cpp:
```diff
@@ -0,0 +1,236 @@
+    pointer_data.w_sat             = w_body_est_fp->get();
+    pointer_data.b                 = b_body_rd_fp->get();
+
+    if (lin::any(!(lin::isfinite(pointer_data.primary_desired) && lin::isfinite(pointer_data.primary_current))))
```
I know I wrote these initially but reading through them again they're a little hard to follow... This may be cleaner:
```suggestion
    if ( !lin::all(lin::isfinite(pointer_data.primary_desired)) || !lin::all(lin::isfinite(pointer_data.primary_current)))
```
totally up to you though.

kkrol27 commented on src/fsw/FCCode/AttitudeController.cpp:
```diff
@@ -0,0 +1,236 @@
+    if (lin::any(!(lin::isfinite(pointer_data.primary_desired) && lin::isfinite(pointer_data.primary_current))))
+        return;
+
+    if (lin::any(!(lin::isfinite(pointer_data.secondary_desired) && lin::isfinite(pointer_data.secondary_current))))
```
Same here.

kkrol27 commented on src/fsw/FCCode/AttitudeController.cpp:
```diff
@@ -0,0 +1,236 @@
+         * We'll autonomously calculate a pointing objective and then control to
+         * it when in standby or docking.
+         */
+        case adcs_state_t::point_standby:
+        case adcs_state_t::point_docking:
+            default_actuator_commands();
+            default_pointing_objectives();
+            transfer_internal_to_output_vectors();
+            calculate_pointing_objectives();
```
This is a little confusing here that sometimes we copy the values into the output vectors from the switch statement and other times we leave it up to the controller. Not sure if there is a cleaner way of handling that.

kkrol27 reviewed (commented):
All in all looks super clean. I have one concern though about how we're handling the outputs from the control task. I was under the impression the attitude controller would always output actuation suggestions and the `ADCSCommander` would read those in and conditionally copy them into the pre-existing `adcs_cmd.*` fields. That way the controller could signal with NaNs that it doesn't have a suggestion and then the commander would be responsible for actually controlling zero torque or whatever we think is best.

In short I guess what I'm proposing is keep the `adcs_cmd.*` fields completely in the commander and then adding two fields for MTR and wheel commands (`attitude.*`) that the commander would read in.

tanishqaggarwal commented on test/test_fsw_attitude_controller/test_attitude_controller.cpp:
```diff
@@ -0,0 +1,297 @@
+    std::cout << lin::transpose(tf.pointer_vec1_desired_fp->get());
+    std::cout << lin::transpose(tf.pointer_vec2_desired_fp->get());
+
+    // all things related to the pointing objectives should be NaN
```
This comment seems false

tanishqaggarwal reviewed (approved):
Looks solid.

kkrol27 commented:
@shihaocao has the scope of this PR changed to getting a detumble HOOTL working as well?

shihaocao commented:
Main changes since last review:

- In order to remove the temporary data structures in controller, I made the controller publish its own suggested commands
- Then ADCSCommander will read these commands, and conditionally forward them to the controller if desired
- I also restuctured ADCSCommander to hopefully be easier to understand

- Updated unit tests as required for both ADCSCommander and AttitudeController

tanishqaggarwal commented on ptest/cases/utils.py:
```diff
@@ -227,7 +227,7 @@ def setup_boot(self):
             # For now, set this to 1 since the value doesn't matter (we force the satellite
```
Does this stuff work now?

tanishqaggarwal commented on src/fsw/FCCode/ADCSCommander.cpp:
```diff
@@ -94,64 +106,65 @@ void ADCSCommander::execute() {
-}
-void ADCSCommander::dispatch_limited(){
-    // TODO: set to desired mag_moment for pointing strat
+        /** We do nothing here since ptest/ground will write into the adcs_cmd fields
```
In my head, the name point_manual implies the _pointing strategy_ is being set manually, not the individual wheel/MTR commands. In that case it should have the same logic as point_standby. I feel like the name "manual" is better

tanishqaggarwal replied:
If you do change this, please also change the corresponding enums table in `ptest/cases/utils.py`

tanishqaggarwal commented on src/fsw/FCCode/ADCSCommander.hpp:
```diff
@@ -53,15 +59,14 @@ class ADCSCommander : public TimedControlTask<void> {
-    void dispatch_manual();     
-    void dispatch_standby();    
-    void dispatch_docking();    
+    template<typename T, size_t N>
```
This is hacky, wondering if we can move to some place better.

tanishqaggarwal commented on telemetry:
```diff
@@ -1713,6 +1713,22 @@
             "type": "unsigned int",
             "writable": true
         },
+        "pointer.mtr_cmd": {
```
It's weird to me that the maximum torque of the magnetorquers is greater than the maximum torque of the reaction wheels. Is this correct?

tanishqaggarwal replied:
Resolution: the number here is the commanded magnetic moment, not the torque. The torque is found via u x B and B ~ 10^-5, so the torque is lower than the reaction wheel torque.
